### PR TITLE
feat: Allow COPY FROM/INTO different storage services

### LINKED ISSUE #6103: Support other stage storage types

### PR DESCRIPTION
**Pull request by Xuanwo**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This PR will close #6103 #6104

This PR adds endpoint URL support for COPY, user can specify the endpoint URL like the following:

```sql
COPY INTO mytable
    FROM 's3://mybucket/data.csv'
    CONNECTION = (
        ENDPOINT_URL = 'http://127.0.0.1:9900'
    )
    FILE_FORMAT = (
        type = 'CSV'
        field_delimiter = ','
        record_delimiter = '\n'
        skip_header = 1
    )
    size_limit=10;
```

Within this PR, we introduced all storage backends support for databend. Now, we can copy data from `azblob`, `hdfs`, `fs`, and so on:

From azblob:

```sql
COPY INTO mytable
    FROM 'azblob://mybucket/data.csv'
    CONNECTION = (
        ENDPOINT_URL = 'http://127.0.0.1:9900'
    )
    FILE_FORMAT = (
        type = 'CSV'
        field_delimiter = ','
        record_delimiter = '\n'
        skip_header = 1
    )
    size_limit=10;
```

To avoid users accessor internal network, we added config for databend:

```toml
[storage]
allow_insecure = false
```

Users can only use an endpoint that starts with `https://` unless `allow_insecure` has been enabled during deployment.

Also, in this PR, we unify all connection-related options into `CONNECTION`:

```sql
COPY INTO mytable
    FROM 'azblob://mybucket/data.csv'
    CONNECTION = (
        ENDPOINT_URL = 'http://127.0.0.1:9900'
        ACCESS_KEY_ID = 'access_key_id'
        SECRET_ACCESS_KEY = 'secret_access_key'
    )
```

`CREDENTIALS` and `ENCRYPTION` are still supported for backward compatibility.

## Remaining Work

- `CREATE STAGE` is not updated, we will adapt changes within the following PRs (addressed in https://github.com/datafuselabs/databend/issues/6580)
- Add tests for copy from azblob.

### CODE REVIEW

File: query/tests/it/configs.rs [hunk 1]
```diff
@@ -91,6 +91,7 @@ rpc_tls_meta_service_domain_name = "localhost"
 [storage]
 type = "fs"
 num_cpus = 0
+allow_insecure = false
```
[the remark this answers is not in the document]
Comment (Xuanwo) [reply query/tests/it/configs.rs:94]: > `allow_insecure` default is false, why explicitly set it here?

This is the default config value test, all values will be listed explicitly.

**Comment by BohuTANG** [comment]
The `Dev MacOS / test_stateless_cluster_macos` failure is not related to this PR

**Comment by Xuanwo** [comment]
@BohuTANG raised some security concern, here are my test results:

What happens if I start `databend-query` with a minio running in `127.0.0.0：9900`？

Copy via `http://127.0.0.1:9900`

```sql
MySQL [(none)]> copy into x from 's3://testbucket/' connection = (endpoint_url='http://127.0.0.1:9900' access_key_id='minioadmin' secret_access_key='minioadmin')  PATTERN = 'ontime_200.csv$' FILE_FORMAT = (type = CSV field_delimiter = ','  record_delimiter = '');
ERROR 1105 (HY000): Code: 3903, displayText = copy from insecure storage is not allowed.
```

Copy via `https://127.0.0.1:9900`

```sql
MySQL [(none)]> copy into x from 's3://testbucket/' connection = (endpoint_url='https://127.0.0.1:9900' access_key_id='xxxx' secret_access_key='yyyy')  PATTERN = 'ontime_200.csv$' FILE_FORMAT = (type = CSV field_delimiter = ','  record_delimiter = '');
ERROR 1105 (HY000): Code: 4000, displayText = other error (backend error: (context: {"bucket": "testbucket"}, source: sending request: https://127.0.0.1:9900/testbucket: hyper::Error(Connect, Ssl(Error { code: ErrorCode(1), cause: Some(Ssl(ErrorStack([Error { code: 336130315, library: "SSL routines", function: "ssl3_get_record", reason: "wrong version number", file: "ssl/record/ssl3_record.c", line: 331 }]))) }, X509VerifyResult { code: 0, error: "ok" })))).
```

**Review by sundy-li — APPROVED** [review]

**Review by junnplus — APPROVED** [review]

**Comment by BohuTANG** [comment]
This is the enhanced version of `COPY`, we are making no silos of data on the cloud, let's document cc @soyeric128

**Comment by soyeric128** [comment]
I have a few questions about this new feature:

- The existing document already says users can optionally specify an endpoint URL for s3 external located files. See https://databend.rs/doc/reference/sql/dml/dml-copy-into-table#externallocation-for-amazon-s3-like. So now, after being included in the CONNECTION part, is it still optional, or a must now?
- Same question for ACCESS_KEY_ID & SECRET_ACCESS_KEY. The existing document indicates they are optional as well. How about now?

**Comment by Xuanwo** [comment]
> I have a few questions about this new feature:

I replied in https://github.com/datafuselabs/databend/issues/6615

**Pull request merged**